(draft) tts: Orpheus support

jsrgb opened the pull request:
A rough draft of SNAC conversion to .gguf with `convert_hf_to_gguf.py`, will add support for this model incrementally, otherwise the PR may be helpful to others.

The upstream `config.json` (https://huggingface.co/hubertsiuzdak/snac_24khz/resolve/main/config.json) does not contain the following which I added manually:
```
    "n_layers": 4,
    "architectures": ["SNACDec"],
    "decoder_channel_dims": [1024, 512, 256, 128, 64]
```
This gets conversion working, but will need to make some tweaks to infer this information from the weights and avoid changes to `config.json`. Next steps are to try decoding with some sample orpheus tokens.

reference issue: #12476

jsrgb committed:
(draft) tts: Orpheus support

Working on each part incrementally, added a rough draft for SNAC convertion to .gguf

ggerganov commented:
Good job. Let us know if you have any questions. You might also find some answers by looking at the commits of the OuteTTS PR: #10784

jsrgb committed:
Scaffolding for snake activation fn

SNAC uses the snake activation function. Added scaffolding to include
`GGML_OP_SNAKE` as a new op. Should this be a unary op?

The SNAC decoder uses noise blocks to enhance outputs, its optional,
so omitting it for now until the model is integrated e2e.

Next steps: write the `llm_graph_context` for SNAC

jsrgb committed:
(rebase me) snac graph, tensor layour and creation

now, integrate the LM (seems straightforward it's llama3), rewrite/extend/add
to tts.cpp, then fix bugs and optimize.

jamorphy commented:
I'm still working on this PR. Orpheus is outputting tokens fine - now ironing out issues in the SNAC graph. I'm aiming to get a reviewable PR out in a few days.

jsrgb committed:
codes to graph build

jsrgb committed:
Merge pull request #1 from jamorphy/wip-orpheus-tts

WIP orpheus tts

jamorphy commented on src/llama-arch.cpp:
```diff
@@ -1391,6 +1392,55 @@ static const std::map<llm_arch, std::map<llm_tensor, const char *>> LLM_TENSOR_N
         },
     },
+    {
+        LLM_ARCH_SNAC_DEC,
```
fix this. do we really need to create a new tensor type for every sub block and res unit

jsrgb committed:
cleanup

jsrgb committed:
Merge pull request #2 from jamorphy/wip-orpheus-tts

cleanup

jsrgb committed:
A forward pass

Run forward passes with dummy codes. Output tensor shapes (raw audio samples) seem to match expected shape given number of input frames. Attempts with Orpheus to be done soon.

The gguf used in this commit is at: https://huggingface.co/jamorphy/snac-fwd-pass-devel-gguf

jsrgb committed:
Merge pull request #3 from jamorphy/wip-orpheus-tts

A forward pass

jsrgb committed:
lazy inputs for snac codes tensors instead of cpu buffers

jamorphy commented:
Running into speed troubles during graph compute, likely due to some operations being done on the CPU. Is there a profiling tool for the compute graph or something similar? For now I'm logging in `ggml_backend_sched_graph_compute_async`, but was wondering if there's something better I may have missed

ggerganov commented:
For profiling individual ops, you can use:

```bash
# profile the GGML_OP_ADD (see the source code for the defined perf tests)
./bin/test-backend-ops -o ADD perf
```

Although for now I think you can just focus on correctness and leave the performance optimizations for later.

ngxson commented on src/llama-model.cpp:
```diff
@@ -11597,6 +11742,196 @@ struct llm_build_wavtokenizer_dec : public llm_graph_context {
+                    cur = ggml_snake(ctx0, cur, alpha);
+
+                    ggml_tensor * w = layer.decoder_blocks[1].up_weight;
+                    ggml_tensor * s = ggml_cpy(ctx0, layer.decoder_blocks[1].up_scale,
```
Out of curiosity, why we need to copy the tensor here?

ngxson replied:
In this case, you can use `ggml_cast`

ngxson reviewed (commented):
Re. your question about performance, some conv_1d may not be available on all backends, so I suspect there are many copies back and forth between CPU and GPU.

My [kyutai-mimi.cpp](https://github.com/ngxson/ggml-easy/blob/master/demo/kyutai-mimi.cpp) implementation does run much faster on CPU compared to GPU because of this. And btw I usually experiment things on ggml-easy first, as there are many debugging tool there, then copy the cgraph over llama.cpp once I'm happy with it. Probably this could help you do faster experiments on ggml.

logikstate commented:
> Been some time since I looked at this but I'll check out ggml-easy.

has somebody already done this? https://github.com/foldl/chatllm.cpp/blob/master/models/orpheus.cpp

andreolf commented:
@jsrgb are you still working on this? If not, I'd like to pick it up and carry it to merge with
credit to you — rebase on `master` and finish the SNAC decode-graph correctness against the
TTS.cpp reference (mmwillet/TTS.cpp #90), then perf. Happy to co-author and keep your commits.
Let me know if you'd rather keep driving it yourself.
